src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -243,6 +244,14 @@ public static int PopCount(uint value)
                 return (int)Popcnt.PopCount(value);
             }
 
+            if (AdvSimd.IsSupported && AdvSimd.Arm64.IsSupported)
+            {
+                // PopCount works on vector so convert input value to vector first.
+                Vector64<uint> input = Vector64.CreateScalar(value);
+                Vector64<byte> aggregated = AdvSimd.Arm64.AddAcross(AdvSimd.PopCount(input.AsByte()));
```

Comment:
It is technically possible to implement this by requiring `AdvSimd.IsSupported` only:

```suggestion
var bytes = Vector64.CreateScalarUnsafe(value).AsByte(); // <b0,b1,b2,b3,*,*,*,*>
var popCounts = AdvSimd.PopCount(bytes) // <p0,p1,p2,p3,*,*,*,*>
var aggregated1 = AdvSimd.AddPairwise(popCounts, Vector64.Zero); // <p0+p1,p2+p3,*,*,0,0,0,0>
var aggregated2 = AdvSimd.AddPairwise(aggregated1, Vector64.Zero); // <p0+p1+p2+p3,*,0,0,0,0,0,0>
return AdvSimd.Extract(aggregated2, 0);
```

While it looks like it's going to be more expensive - you would actually save one instruction on using CreateScalarUnsafe instead of CreateScalar. 


You can further eliminate requirement on Vector64.Zero by using AdvSimd.AddPairwiseWidening
For example,

```suggestion
var bytes = Vector64.CreateScalarUnsafe(value).AsByte(); // <b0,b1,b2,b3,*,*,*,*>
var popCounts = AdvSimd.PopCount(bytes) // <p0,p1,p2,p3,*,*,*,*>
var aggregated1= AdvSimd.AddPairwiseWidening(popCounts); // <p0+p1,p2+p3,*,*>
var aggregated2 = AdvSimd.AddPairwiseWidening(aggregated1); // <p0+p1+p2+p3,*>
return AdvSimd.Extract(aggregated2, 0);
```

However, I don't which variant is going to be faster.
cc @TamarChristinaArm

Reply:
We definitely don't want to do `CreateScalarUnsafe()` because the current implementation uses `movi` (I think we spoke offline about it when implementing `CreateScalarUnsafe`) which moves the immediate into every lane.
E.g. `Vector64.CreateScalarUnsafe((byte)23)` generates `movi    v16.2s, #0x17` and hence the vector ends up being `<23, 23>`.

Reply:
If we use AddPairwise or AddPairwiseWidening it shouldn't matter.

Reply:
@TamarChristinaArm  can confirm this, but as per [this docs](https://static.docs.arm.com/epm128372/30/arm_cortex_a55_software_optimization_guide_v3.pdf) the latency of `AddAcross` and AddPairwiseWidening` is same. So I think it will be optimal to use `AddAcross` once rather than `AddPairwiseWidening` twice.

Reply:
In your example, why is it `Vector64.CreateScalarUnsafe((byte)23)` when it should be `Vector64.CreateScalarUnsafe((uint)23)`?

Then
Vector64.CreateScalarUnsafe((uint)23) -> <23, 23>
AdvSimd.PopCount(<23, 23>) -> <3, 1, 0, 0, 3, 1, 0, 0>
AdvSimd.AddPairwiseWidening(<3, 1, 0, 0, 3, 1, 0, 0>) -> <4, 0, 4, 0>
AdvSimd.AddPairwiseWidening(<4, 0, 4, 0>) -> <4, 4>
AdvSimd.Extract(<4, 4>, 0) -> 4

Reply:
> In your example, why is it Vector64.CreateScalarUnsafe((byte)23) when it should be Vector64.CreateScalarUnsafe((uint)23)?

Sorry, that was a typo. It should have been `Vector64.CreateScalarUnsafe((uint)23)`.

Reply:
Current implementation:

```asm
G_M24818_IG02:
        4E081C10          mov     v16.d[0], x0
        0E205A10          cnt     v16.8b, v16.8b
        0E31BA10          addv    b16, v16.8b
        0E013E00          umov    w0, v16.b[0]
                                                ;; bbWeight=1    PerfScore 5.00
```

Your proposal:

```asm
G_M59888_IG02:
        9E630010          ucvtf   d16, x0
        1E624210          fcvt    s16, d16
        0E205A10          cnt     v16.8b, v16.8b
        0F000411          movi    v17.2s, #0x00
        0E31BE10          addp    v16.8b, v16.8b, v17.8b
        0F000411          movi    v17.2s, #0x00
        0E31BE10          addp    v16.8b, v16.8b, v17.8b
        0E013E00          umov    w0, v16.b[0]
                                                ;; bbWeight=1    PerfScore 14.00
```
Current implementation: 1.5408738 secs.
Your proposal : 1.539505 secs.

I don't see much difference between the two but we generate more instructions with `AddPairwise`.

Reply:
Above example was for `ulong`. But `uint` one sees benefit with your suggestion.

Current implementation:

```asm
G_M36043_IG02:
        0F000410          movi    v16.2s, #0x00
        4E041C10          ins     v16.s[0], w0
        FD000FB0          str     d16, [fp,#24]
        FD400FB0          ldr     d16, [fp,#24]
        0E205A10          cnt     v16.8b, v16.8b
        0E31BA10          addv    b16, v16.8b
        0E013E00          umov    w0, v16.b[0]
                                                ;; bbWeight=1    PerfScore 8.50
```

Your proposal:
```asm
G_M44169_IG02:
        4E041C10          ins     v16.s[0], w0
        0E205A10          cnt     v16.8b, v16.8b
        0F000411          movi    v17.2s, #0x00
        0E31BE10          addp    v16.8b, v16.8b, v17.8b
        0F000411          movi    v17.2s, #0x00
        0E31BE10          addp    v16.8b, v16.8b, v17.8b
        0E013E00          umov    w0, v16.b[0]
                                                ;; bbWeight=1    PerfScore 8.00

Current implementation: 3.6687005 secs.
Your proposal : 1.5394875 secs.

With my implementation, we spend time spilling to memory for converting to `Vector64<byte>`.

Reply:
@kunalspathak Can you please also try `uaddpl` - AddPairwiseWidening? It should also eliminate this 
```asm
 movi    v17.2s, #0x00
```

Reply:
Actually, there is a better idea, you can try
```c#
var bytes = Vector64.CreateScalarUnsafe(value).AsByte();
var popCounts = AdvSimd.PopCount(bytes);
var aggregated1 = AdvSimd.AddPairwise(popCounts, popCounts);
var aggregated2 = AdvSimd.AddPairwise(aggregated1, aggregated1);
return AdvSimd.Extract(aggregated2, 0);
```
the jit should optimize this down to
```asm
ins     v16.s[0], w0
cnt     v16.8b, v16.8b
addp    v16.8b, v16.8b, v16.8b
addp    v16.8b, v16.8b, v16.8b
umov    w0, v16.b[0]
```

Reply:
Have we looked at what clang generates for `__builtin_popcount`?

Reply:
> So i think I will go ahead with your suggestion unless @TamarChristinaArm wants to add something.
Ah sorry, different timezones :)

> @TamarChristinaArm can confirm this, but as per this docs the latency of AddAcross and AddPairwiseWidening is same. So I think it will be optimal to use AddAcross once rather than AddPairwiseWidening twice.

Yes the `addv` approach is better.  the `addp` while having a lower latency needs data from the previous one, so even though it can go into both pipelines it doesn't matter since it still has to wait for the previous one.

> Current implementation: 3.6687005 secs.
> Your proposal : 1.5394875 secs.

That's because it's not comparing apples to apples

```
G_M36043_IG02:
        0F000410          movi    v16.2s, #0x00
        4E041C10          ins     v16.s[0], w0
        FD000FB0          str     d16, [fp,#24]
        FD400FB0          ldr     d16, [fp,#24]
        0E205A10          cnt     v16.8b, v16.8b
        0E31BA10          addv    b16, v16.8b
        0E013E00          umov    w0, v16.b[0]
```

this one has a load and store to memory, your slow down is coming from that. while

```
G_M44169_IG02:
        4E041C10          ins     v16.s[0], w0
        0E205A10          cnt     v16.8b, v16.8b
        0F000411          movi    v17.2s, #0x00
        0E31BE10          addp    v16.8b, v16.8b, v17.8b
        0F000411          movi    v17.2s, #0x00
        0E31BE10          addp    v16.8b, v16.8b, v17.8b
        0E013E00          umov    w0, v16.b[0]
```
managed to avoid it.

Your optimal sequence here is

```
G_M36043_IG02:
        4E041C10          fmov     s16, w0
        0E205A10          cnt     v16.8b, v16.8b
        0E31BA10          addv    b16, v16.8b
        0E013E00          umov    w0, v16.b[0]
```

by using `fmov` instead of `mov` you can avoid having to clear the register first. and if you also avoid the memory access it should beat the operation with the `addp` above.

Reply:
Also there's a lot of variance in the numbers, how'd it go from `With optimization: 1.8343483 secs.` to `Current implementation: 3.6687005 secs.` ?

Reply:
> __builtin_popcount

It uses `fmov` and `addv`. See https://godbolt.org/z/uFN5dv

> Also there's a lot of variance in the numbers, how'd it go from With optimization: 1.8343483 secs. to Current implementation: 3.6687005 secs. ?

I double checked the time difference and as you pointed out it is coming from memory access. 

<details>
<summary>Here is the benchmark code that I am using.</summary>

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static int Current(uint value) {	
	int sum = 0;
	var stopWatch = new Stopwatch();
	stopWatch.Start();
	for (int i = 0; i < 1000 * 1000 * 2000; i++)
	{
		if (AdvSimd.IsSupported)
		{
			// Create vector from input value
			Vector64<uint> input = Vector64.CreateScalar(value);
			Vector64<byte> aggregated = AdvSimd.Arm64.AddAcross(AdvSimd.PopCount(input.AsByte()));
			sum += (int)AdvSimd.Extract(aggregated, 0);
		}
	}
	stopWatch.Stop();
	Console.WriteLine("Time taken Current: " + stopWatch.Elapsed.TotalSeconds);
	return sum;
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static int Proposed(uint value) {	
	int sum = 0;
	var stopWatch = new Stopwatch();
	stopWatch.Start();
	for (int i = 0; i < 1000 * 1000 * 2000; i++)
	{
		if (AdvSimd.IsSupported)
		{
			var bytes = Vector64.CreateScalarUnsafe(value).AsByte(); // <b0,b1,b2,b3,*,*,*,*>
			var popCounts = AdvSimd.PopCount(bytes); // <p0,p1,p2,p3,*,*,*,*>
			var aggregated1 = AdvSimd.AddPairwise(popCounts, Vector64<byte>.Zero); // <p0+p1,p2+p3,*,*,0,0,0,0>
			var aggregated2 = AdvSimd.AddPairwise(aggregated1, Vector64<byte>.Zero); // <p0+p1+p2+p3,*,0,0,0,0,0,0>
			sum +=  AdvSimd.Extract(aggregated2, 0);
		}
	}
	stopWatch.Stop();
	Console.WriteLine("Time taken Proposed: " + stopWatch.Elapsed.TotalSeconds);
	return sum;
}
```

</details>


Even I was surprised with the memory access in my implementation which is biggest contributor to the time difference. It is interesting to see that the reason we do the memory access is because for [Vector64<uint> CreateScalar(uint)](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector64.cs#L644) we retype the return type as `double`. This happens for other variants of `CreateScalar()` like byte, short, etc.

```asm
impFixupStructReturnType: retyping
               [000039] ------------              *  HWINTRINSIC simd8  uint Insert
               [000038] ------------              \--*  LIST      void  
               [000034] ------------                 +--*  HWINTRINSIC simd8  uint get_Zero
               [000037] ------------                 \--*  LIST      void  
               [000035] ------------                    +--*  CNS_INT   int    0
               [000036] ------------                    \--*  LIST      void  
               [000009] ------------                       \--*  LCL_VAR   int    V00 arg0         

impFixupStructReturnType: result of retyping is
               [000039] ------------              *  HWINTRINSIC double uint Insert
               [000038] ------------              \--*  LIST      void  
               [000034] ------------                 +--*  HWINTRINSIC simd8  uint get_Zero
               [000037] ------------                 \--*  LIST      void  
               [000035] ------------                    +--*  CNS_INT   int    0
               [000036] ------------                    \--*  LIST      void  
               [000009] ------------                       \--*  LCL_VAR   int    V00 arg0         


               [000041] -A----------              *  ASG       double
               [000040] D------N----              +--*  LCL_VAR   double V05 tmp2         
               [000039] ------------              \--*  HWINTRINSIC double uint Insert
               [000038] ------------                 \--*  LIST      void  
               [000034] ------------                    +--*  HWINTRINSIC simd8  uint get_Zero
               [000037] ------------                    \--*  LIST      void  
               [000035] ------------                       +--*  CNS_INT   int    0
               [000036] ------------                       \--*  LIST      void  
               [000009] ------------                          \--*  LCL_VAR   int    V00 arg0  

```
I am not sure if this is expected or if we can improve to not do it. @CarolEidt , is it possible to avoid doing it because I can see lot of places where result of `CreateScalar()` being used and we would spill the contents to memory before using the result. So basically my implementation is slower because of this issue. 

However the retyping doesn't happen for `ulong` case and hence there is no difference in my current implementation and what Egor proposed. So I think i will stick to my implementation for `ulong`. For `uint` because of the current shortcoming with `CreateScalar()` I don't see a way to create a vector that will just have value of `uint` so I can do `AddAcross` instead of `AddPairwise`.  So I am thinking of changing `uint` implementation to just call `ulong` by doing `PopCount((ulong)value);` because the generated code of `ulong` is efficient and matches what gcc generates. That adds 1 extra `mov w0, w0` to zero extend and then use x0 for further operations, but still slicker and don't incur any performance penalty.

Reply:
Is that potentially due to an assumption that `TYP_SIMD8` means `System.Numerics.Vector2` (as before HWIntrincs on ARM64, that is all it could be)?

Reply:
> It uses fmov and addv. See https://godbolt.org/z/uFN5dv

You probably want `__builtin_popcountl` for the `long` case otherwise the result is wrong.

> So I am thinking of changing uint implementation to just call ulong implementation by doing PopCount((ulong)value); because the generated code of ulong is efficient and matches what gcc generates.

Yeah I agree that you should stick to the `addv`. 

> That adds 1 extra mov w0, w0 to zero extend and then use x0 for further operations, but still slicker and don't incur any performance penalty.

If the extend the same problem as with `CreateScalar`? i.e. it doesn't know that only the bottom 32-bits will be used so it doesn't need to extend? To be fair we did only clean that up in GCC 9.

Reply:
> Is that potentially due to an assumption that TYP_SIMD8 means System.Numerics.Vector2, 

We correctly determine it to be intrinsic API and return [vector size as 8 bytes](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/class.cpp#L1204), but then over [here](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/jitinterface.cpp#L346) `eeType=ELEMENT_TYPE_R8 (0n13)` maps to `CORINFO_TYPE_DOUBLE`. I am not much familiar with this method, but I believe somewhere after we get that value (probably inside [getPrimitiveTypeForStruct](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/compiler.cpp#L544) or [getReturnTypeForStruct](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/compiler.cpp#L971) we should see if this is simd type and map to one of `TYP_SIMD8`.

Callstack just for the reference:
```
00 CoreCLR!CEEInfo::asCorInfoType
01 CoreCLR!CEEInfo::getHFAType
02 clrjit!Compiler::GetHfaType
03 clrjit!Compiler::getPrimitiveTypeForStruct
04 clrjit!Compiler::getReturnTypeForStruct
05 clrjit!Compiler::IsMultiRegReturnedType
06 clrjit!Compiler::impFixupStructReturnType
07 clrjit!Compiler::impReturnInstruction
08 clrjit!Compiler::impImportBlockCode
09 clrjit!`Compiler::impImportBlock'::`15'::__Body::Run
0a clrjit!Compiler::impImportBlock
0b clrjit!Compiler::impImport
0c clrjit!Compiler::fgImport

```

> You probably want __builtin_popcountl for the long case otherwise the result is wrong.

Ah ok. Updated here: https://godbolt.org/z/PbT_RP

> Yeah I agree that you should stick to the addv.

Thanks, will send out an update.

> If the extend the same problem as with CreateScalar?

Not sure. I assume it doesn't take in account the fact that it needs only 32-bits.

Reply:
> for Vector64 CreateScalar(uint) we retype the return type as double.

I suspect that this is another fallout from the fact that `GetHfaType` returns `TYP_DOUBLE` for an HVA of 64-bit vectors, since ABI-wise an HVA of 64-bit vectors is passed the same as an HFA of doubles. However, I believe that a `Vector64<ulong>` should be handled exactly the same as any other `Vector64<T>` so that seems a bit concerning.

Reply:
@kunalspathak could you open an issue for the fact that these vector types are handled differently (and specifically for the non-ulong case they are pessimized?)

Reply:
Thanks Carol. Opened https://github.com/dotnet/runtime/issues/35976